### DISCUSSION TITLE
Dominance and GxE

### DISCUSSION THREAD

**Discussion by TongYinCast** [gaynorr/AlphaSimR#104] [Q&A]
Dear AlphaSimR authors,

My name is Tong Yin a postdoc at Giessen University in Germany. Recently, my colleagues and I are working on a simulation study, using AlphaSimR. Both dominance and GxE effects are interesting for us. But I don’t really get how the dominance effects and the dominance deviations are sampled and calculated, even after reading the vignette “Traits”. If I understand correctly, the dominance effects are sampled through dominance degrees and the absolute values of additive effects, which means that if I only set DDmean to e.g., 0.5, then I would get only positive dominance effects for each QTL (since the default VarDD = 0), resulting in positive dd for the animals in the initial population (G0), right? After repeating the initial populations 100 times, I got minimum dd from -2.5 to -3.5… I am just wondering if I did something wrong in the script. Please find my script below.

nCows = 5000
nChr = 10
nQtl = 1000
nSnp = 5000

FOUNDERPOP = runMacs(nInd= nCows,
                     nChr=nChr,
                     segSites=nQtl+nSnp,
                     species = "CATTLE",
                     nThreads = 5)
summ.ad <- array(0, c(100,6))
varall <- array(0,c(100,4))
for (i in 1:100){
  SP = SimParam$new(FOUNDERPOP)
  SP$resetPed()
  SP$setTrackPed(TRUE) 
  SP$setSexes("yes_rand")
  SP$addSnpChip(nSnpPerChr=nSnp)
  #additive + dominance effects
  SP$addTraitAD(nQtlPerChr=nQtl,mean=0,var=1,meanDD = 0.25) #, varDD = 0.09
  SP$setVarE(h2 = 0.3)
  
  initPop = newPop(FOUNDERPOP, simParam = SP)
  varall[i,] <- c(varA(initPop), varD(initPop), varG(initPop),varP(initPop))
  summ.ad[i,]<-c(min(bv(initPop)),max(bv(initPop)),mean(bv(initPop)),min(dd(initPop)),max(dd(initPop)),mean(dd(initPop)))` 
}
mean(varall[,2])
summary(varall[,1]/varall[,4])
summary(varall[,2]/varall[,4])
summary(varall[,3]/varall[,4])
summary(summ.ad)

One more question is related to GxE interactions. It seems the environmental covariate was random selected like this: qnorm(runif(1)). It is not clear here if you generate one p-value from the uniform distribution when using setPheno or a vector of p-values (depends on number of animals with phenotypes) in one replicate. In our simulation study, we want to simulate two breeds with the consideration of dominance effects and GxE, when we cross the two breeds. Initially, the two breeds were sampled from the same founder population, and then selected separately on different traits for several generations. So what I would expect is a re-ranking of TBV when animals are transported from one population to other population. I get lost when I apply the GxE argument in AlphaSimR. Do you recommend using the GxE argument to realize the re-ranking? Or modification of TBV or simulation of two genetic correlated traits (the same trait in two environments) with pre-defined accuracies/genetic correlations is a better option?

In addition, is it possible to extract additive and dominance effects for each QTL in AlphaSimR? 

Best regards,
Tong

**Comment by TongYinCast** [comment]
QTLgeno <- pullQtlGeno(initPop)
addef <- rnorm(ncol(QTLgeno))
bv.ind <- QTLgeno%*%addef
sbv.ind <- 1/sd(bv.ind)*bv.ind

dom.deg <- rnorm(ncol(QTLgeno), mean = 0.2, sd = 0.3)
domef <- dom.deg*abs(addef)
QTLgeno.dom <- QTLgeno
QTLgeno.dom[QTLgeno.dom==2] <- 0
dd.ind <- QTLgeno.dom%*%domef
sdd.ind <- 1/sd(bv.ind)*dd.ind
summary(sdd.ind-mean(sdd.ind))

I think I got the reason for the negative dd in the initial population, because the mean of dd is subtracted from sdd to keep the mean of gv equals to 0 in G0 :) 

But would be nice, if you can explain me the GxE argument, thanks

**Reply by gaynorr** [reply]
The same p-value for the GxE effect gets applied to all individuals in the population. The underlying assumption is that all individuals in the population are being exposed to the same environment. 

I don't include the GxE effects in the breeding value and dominance deviation calculations. For these calculations, I only assume the average environments (p=0.5) which effectively zeros out the GxE effects due to the covariate equaling zero. I mainly did this because it was the simplest thing to do and it wasn't clear to me that any other option would be better. I'm open to suggestions for a different way of doing it.

**Comment by gregorgorjanc** [comment]
> If I understand correctly, the dominance effects are sampled through dominance degrees and the absolute values of additive effects, which means that if I only set DDmean to e.g., 0.5, then I would get only positive dominance effects for each QTL (since the default VarDD = 0), resulting in positive dd for the animals in the initial population (G0), right? After repeating the initial populations 100 times, I got minimum dd from -2.5 to -3.5…

You are confusing biological dominance effects (`d`, which we sample as `d = abs(a) * delta`, where `a` is biological additive effect of alleles and `delta ~ N(meanDD, varDD)`) with statistical dominance deviations (`dd`). The latter depend on allele frequencies etc. so you can indeed get a range of values. Have a look in the the `SP$traits` for these biological effects.

@gaynorr can add more;)

**Reply by TongYinCast** [reply]
Thanks for your reply. Then, the equation behind the bv function should be equation 6 from the paper by Faux et al., 2016, right? The dd is only the difference between gv and sum of bv + mu. I thought they were calculated using equation 4 and 5 in the vignette "Traits". Thanks for the corretion.

**Reply by gaynorr** [reply]
I recommend avoiding using Faux et al. as a reference for AlphaSimR, because the latter was completely rewritten. This resulted in quite a lot of changes. With regards to breeding values and dominance deviations, AlphaSim and AlphaSimR are very different. This difference comes from the use of different reference populations. 

AlphaSim used a reference population that is in linkage equilibrium, Hardy-Weinberg equilibrium and has allele frequencies matching the base generation. These assumptions give rise to the formulas you see in the document. I didn't see any utility in using this as a reference population, because the resulting breeding values and dominance deviations are not appropriate for measuring additive and dominance variance in the current population.

AlphaSimR instead uses a reference population that is in linkage equilibrium and has genotype frequencies matching the current population. In other words, I removed the Hardy-Weinberg assumption by using genotype and not allele frequencies. I also made the reference corresponds to the current population, so that additive and dominance variances could be calculated in the current population. By definition, the resulting breeding values and dominance deviations will have a mean of zero. There's not a really simple formula for this calculation, because AlphaSimR potentially models autopolyploids and/or additive-by-additive epistasis.
